src/V3FuncOpt.cpp

```diff
@@ -0,0 +1,184 @@
+//
+//*************************************************************************
+//
+// Copyright 2003-2024 by Wilson Snyder. This program is free software; you
+// can redistribute it and/or modify it under the terms of either the GNU
+// Lesser General Public License Version 3 or the Perl Artistic License
+// Version 2.0.
+// SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0
+//
+//*************************************************************************
+//
+// - Split assignments to wide locations with Concat on the RHS
+//   at word boundaries:
+//    foo = {l, r};
+//   becomes (recursively):
+//    foo[_:_] = r;
+//    foo[_:_] = l;
+//
+//*************************************************************************
+
+#include "V3PchAstMT.h"
+
+#include "V3FuncOpt.h"
+
+#include "V3Global.h"
+#include "V3Stats.h"
+#include "V3ThreadPool.h"
+
+VL_DEFINE_DEBUG_FUNCTIONS;
+
+class FuncOptVisitor final : public VNVisitor {
+    // NODE STATE
+    //  AstNodeAssign::user()     -> bool.  Already checked, safe to split. Omit expensive check.
+
+    // STATE - across all visitors
+    AstCFunc* const m_funcp;  // The function being processes
+
+    // STATE - Statistic tracking
+    VDouble0 m_concatSplits;  // Number of splits in assignments with Concat on RHS
+
+    // True for e.g.: foo = foo >> 1; or foo[foo[0]] = ...;
+    static bool readsLhs(AstNodeAssign* nodep) {
+        // It is expected that the number of vars written on the LHS is very small (should be 1).
+        std::vector<const AstVar*> lhsWrVarps;
+        std::unordered_set<const AstVar*> lhsRdVarps;
+        nodep->lhsp()->foreach([&](const AstVarRef* refp) {
+            if (refp->access().isWriteOrRW()) lhsWrVarps.emplace_back(refp->varp());
+            if (refp->access().isReadOrRW()) lhsRdVarps.emplace(refp->varp());
+        });
+
+        // Common case of 1 variable on the LHS
+        if (lhsWrVarps.size() == 1) {
+            const bool rhsReadsWritten = nodep->rhsp()->exists([&](const AstVarRef* refp) {  //
+                return refp->varp() == lhsWrVarps[0];
+            });
+            return rhsReadsWritten || lhsRdVarps.count(lhsWrVarps[0]);
+        }
+
+        // Generic case - TODO: should this ever be hit at this point?
+        // Check Rhs
+        const bool rhsReadsWritten = nodep->rhsp()->exists([&](const AstVarRef* refp) {
+            for (const AstVar* const wrVarp : lhsWrVarps) {
+                if (refp->varp() == wrVarp) return true;
+            }
+            return false;
+        });
+        if (rhsReadsWritten) return true;
+        // Check Lhs
+        for (const AstVar* const wrVarp : lhsWrVarps) {
+            if (lhsRdVarps.count(wrVarp)) return true;
+        }
+        return false;
+    }
+
+    // METHODS
+    // Split wide assignments with a wide concatenation on the RHS.
+    // Returns true if 'nodep' was deleted
+    bool splitConcat(AstNodeAssign* nodep) {
+        // Only care about concatenations on the right
```

Comment:
```suggestion
    bool splitConcat(AstNodeAssign* nodep) {
        UINFO(9, "splitConcat " << nodep << "\n");
        // Only care about concatenations on the right
```

Reply:
(My first-pass for debugging is printf's - old school.)

Reply:
Done